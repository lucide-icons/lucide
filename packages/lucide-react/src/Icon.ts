import { createElement, forwardRef } from 'react';
import defaultAttributes from './defaultAttributes';
import { IconNode, LucideProps } from './types';
import { mergeClasses } from '@lucide/shared';

interface IconComponentProps extends LucideProps {
  iconNode: IconNode;
}

/**
 * Lucide icon component
 *
 * @component Icon
 * @param {object} props
 * @param {string} props.color - The color of the icon
 * @param {number} props.size - The size of the icon
 * @param {number} props.strokeWidth - The stroke width of the icon
 * @param {boolean} props.absoluteStrokeWidth - Whether to use absolute stroke width
 * @param {string} props.className - The class name of the icon
 * @param {IconNode} props.children - The children of the icon
 * @param {IconNode} props.iconNode - The icon node of the icon
 *
 * @returns {ForwardRefExoticComponent} LucideIcon
 */
const Icon = forwardRef<SVGSVGElement, IconComponentProps>(
  (
    {
      color = 'currentColor',
      size = 24,
      strokeWidth = 2,
      absoluteStrokeWidth,
      className = '',
      children,
      iconNode,
      ...rest
    },
    ref,
  ) => {
    return createElement(
      'svg',
      {
        ref,
        ...defaultAttributes,
        width: size,
        height: size,
        stroke: color,
        strokeWidth: absoluteStrokeWidth ? (Number(strokeWidth) * 24) / Number(size) : strokeWidth,
        className: mergeClasses('lucide', className),
        ...rest,
      },
      [
        ...iconNode.map(([tag, attrs]) => createElement(tag, attrs)),
        ...(Array.isArray(children) ? children : [children]),
      ],
    );
  },
);

export default Icon;
