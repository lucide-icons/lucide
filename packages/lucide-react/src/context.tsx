import { createContext, type ReactNode, useContext, useMemo } from 'react';
import { LucideProps } from './types';

type LucideConfig = {
  size: number;
  fill: string;
  color: string;
  strokeWidth: number;
  absoluteStrokeWidth: boolean;
};

const LucideContext = createContext<LucideProps>({});

type LucideProviderProps = {
  children: ReactNode;
} & Partial<LucideConfig>;

export function LucideProvider({
  children,
  size,
  fill,
  color,
  strokeWidth,
  absoluteStrokeWidth,
}: LucideProviderProps) {
  const value = useMemo(
    () => ({
      size,
      fill,
      color,
      strokeWidth,
      absoluteStrokeWidth,
    }),
    [size, fill, color, strokeWidth, absoluteStrokeWidth],
  );

  return <LucideContext.Provider value={value}>{children}</LucideContext.Provider>;
}

export const useLucideContext = () => useContext(LucideContext);
