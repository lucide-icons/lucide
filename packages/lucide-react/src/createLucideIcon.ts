import { forwardRef, createElement, ReactSVG, SVGProps } from 'react';
import defaultAttributes from './defaultAttributes';

type IconNode = [elementName: keyof ReactSVG, attrs: Record<string, string>][]

export type SVGAttributes = Partial<SVGProps<SVGSVGElement>>

export interface LucideProps extends SVGAttributes {
  size?: string | number
}
/**
 * Converts string to KebabCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = (string: string) => string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

const createLucideIcon = (iconName: string, iconNode: IconNode) => {
  const Component = forwardRef<SVGSVGElement, LucideProps>(
    ({ color = 'currentColor', size = 24, strokeWidth = 2, children, ...rest }, ref) =>
      createElement(
        'svg',
        {
          ref,
          ...defaultAttributes,
          width: size,
          height: size,
          stroke: color,
          strokeWidth,
          className: `lucide lucide-${toKebabCase(iconName)}`,
          ...rest,
        },
        [
          ...iconNode.map(([tag, attrs]) => createElement(tag, attrs)),
          ...(
            (Array.isArray(children) ? children : [children]) || []
          )
        ],
      ),
  );

  Component.displayName = `${iconName}`;

  return Component;
};

export default createLucideIcon
