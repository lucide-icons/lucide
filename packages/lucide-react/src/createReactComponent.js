import { forwardRef, createElement } from 'react';
import PropTypes from 'prop-types';

export default (iconName, [tag, attrs, children]) => {
  const Component = forwardRef(
    ({ color = 'currentColor', size = 24, strokeWidth = 2, ...rest }, ref) =>
      createElement(
        tag,
        {
          ref,
          ...attrs,
          width: size,
          height: size,
          stroke: color,
          strokeWidth,
          ...rest,
        },
        children.map(([childTag, childAttrs]) => createElement(childTag, childAttrs)),
      ),
  );

  Component.propTypes = {
    color: PropTypes.string,
    size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    strokeWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  };

  Component.displayName = `${iconName}`;

  return Component;
};
