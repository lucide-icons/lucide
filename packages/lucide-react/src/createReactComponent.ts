import React, { forwardRef, createElement, ReactSVG, SVGProps } from 'react';
import PropTypes from 'prop-types';
import { IconNode, DEFAULT_ATTRIBUTES } from 'lucide'

export type SVGAttributes = Partial<SVGProps<SVGSVGElement>>

export interface LucideProps extends SVGAttributes {
  size?: string | number
}

/**
 * Converts string to KebabCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = (string: string): string => string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

/**
 * Converts string to CamelCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A Camelized string
 */
export const toCamelCase = (string: string): string => string.replace(/-./g, x=>x[1].toUpperCase());

const defaultAttrs = Object.keys(DEFAULT_ATTRIBUTES).reduce((acc, key)=> {
  acc[toCamelCase(key)] = DEFAULT_ATTRIBUTES[key]
  return acc
}, {} as Record<string,string|number>)

export function createReactComponent (iconName: string, [_tag, DEFAULT_ATTRIBUTES, nodeChildren=[]]: IconNode) {
    const Component = forwardRef<SVGSVGElement, LucideProps>(
    ({ color = 'currentColor', size = 24, strokeWidth = 2, children, ...rest }, ref) =>
      createElement(
        'svg',
        {
          ref,
          ...defaultAttrs,
          width: size,
          height: size,
          stroke: color,
          strokeWidth,
          className: `lucide lucide-${toKebabCase(iconName)}`,
          ...rest,
        },
        [
          ...nodeChildren.map(([tag, attrs], index) => createElement(tag, { key: index, ...attrs })),
          ...(
            (Array.isArray(children) ? children : [children]) || []
          )
        ],
      ),
  );

  Component.propTypes = {
    color: PropTypes.string,
    size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    strokeWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  };

  Component.displayName = `${iconName}`;

  return Component;
};
