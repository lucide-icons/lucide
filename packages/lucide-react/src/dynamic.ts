export {
  default as DynamicIcon,
  iconNames,
  type DynamicIconModule,
  type IconName,
} from './DynamicIcon';
export { default as dynamicIconImports } from './dynamicIconImports';
