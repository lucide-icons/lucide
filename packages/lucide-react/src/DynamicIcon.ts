'use client';

import { createElement, forwardRef, useEffect, useState } from 'react';
import { IconNode, LucideIcon, LucideProps } from './types';
import dynamicIconImports from './dynamicIconImports';
import Icon from './Icon';

export type DynamicIconModule = { default: LucideIcon; __iconNode: IconNode };

export type IconName = keyof typeof dynamicIconImports;

export const iconNames = Object.keys(dynamicIconImports) as Array<IconName>;

interface DynamicIconComponentProps extends LucideProps {
  name: IconName;
  fallback?: () => JSX.Element | null;
}

async function getIconNode(name: IconName) {
  if (!(name in dynamicIconImports)) {
    throw new Error('[lucide-react]: Name in Lucide DynamicIcon not found');
  }

  // TODO: Replace this with a generic iconNode package.
  const icon = (await dynamicIconImports[name]()) as DynamicIconModule;

  return icon.__iconNode;
}

/**
 * Dynamic Lucide icon component
 *
 * @component Icon
 * @param {object} props
 * @param {string} props.color - The color of the icon
 * @param {number} props.size - The size of the icon
 * @param {number} props.strokeWidth - The stroke width of the icon
 * @param {boolean} props.absoluteStrokeWidth - Whether to use absolute stroke width
 * @param {string} props.className - The class name of the icon
 * @param {IconNode} props.children - The children of the icon
 * @param {IconNode} props.iconNode - The icon node of the icon
 *
 * @returns {ForwardRefExoticComponent} LucideIcon
 */
const DynamicIcon = forwardRef<SVGSVGElement, DynamicIconComponentProps>(
  ({ name, fallback: Fallback, ...props }, ref) => {
    const [iconNode, setIconNode] = useState<IconNode>();

    useEffect(() => {
      getIconNode(name)
        .then(setIconNode)
        .catch((error) => {
          console.error(error);
        });
    }, [name]);

    if (iconNode == null) {
      if (Fallback == null) {
        return null;
      }

      return createElement(Fallback);
    }

    return createElement(Icon, {
      ref,
      ...props,
      iconNode,
    });
  },
);

export default DynamicIcon;
