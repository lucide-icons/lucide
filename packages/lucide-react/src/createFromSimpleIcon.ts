import createLucideIcon from './createLucideIcon';
import type { IconNode } from './types';
import type { SimpleIcon } from 'simple-icons';

const createIconNodeFromSimpleIcon = (simpleIcon: SimpleIcon): IconNode => [
  [
    'path',
    {
      d: simpleIcon.path,
      fill: 'currentColor',
    },
  ],
];

/**
 * Converts a simple-icon into a Lucide icon component
 * @param name The name of the icon (will be used as component name)
 * @param simpleIcon The simple-icon object
 * @returns A Lucide icon component
 */
const createFromSimpleIcon = (name: string, simpleIcon: SimpleIcon) =>
  createLucideIcon(name, createIconNodeFromSimpleIcon(simpleIcon));

export default createFromSimpleIcon;
