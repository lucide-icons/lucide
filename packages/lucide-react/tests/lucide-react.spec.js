import React from 'react';
import renderer from 'react-test-renderer';
import { Grid } from '../src/icons'

describe('Using lucide icon components', () => {
  it('should render an component', () => {
    const component = renderer.create(
      <Grid/>,
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should adjust the size, stroke color and stroke width', () => {
    const component = renderer.create(
      <Grid
        size={48}
        stroke="red"
        strokeWidth={4}
      />,
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
})
