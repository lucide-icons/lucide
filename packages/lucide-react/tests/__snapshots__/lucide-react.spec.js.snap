// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Using lucide icon components should adjust the size, stroke color and stroke width 1`] = `
<svg
  fill="none"
  height={48}
  stroke="red"
  strokeLinecap="round"
  strokeLinejoin="round"
  strokeWidth={4}
  viewBox="0 0 24 24"
  width={48}
  xmlns="http://www.w3.org/2000/svg"
>
  <rect
    height="18"
    rx="2"
    ry="2"
    width="18"
    x="3"
    y="3"
  />
  <line
    x1="3"
    x2="21"
    y1="9"
    y2="9"
  />
  <line
    x1="3"
    x2="21"
    y1="15"
    y2="15"
  />
  <line
    x1="9"
    x2="9"
    y1="3"
    y2="21"
  />
  <line
    x1="15"
    x2="15"
    y1="3"
    y2="21"
  />
</svg>
`;

exports[`Using lucide icon components should render an component 1`] = `
<svg
  fill="none"
  height={24}
  stroke="currentColor"
  strokeLinecap="round"
  strokeLinejoin="round"
  strokeWidth={2}
  viewBox="0 0 24 24"
  width={24}
  xmlns="http://www.w3.org/2000/svg"
>
  <rect
    height="18"
    rx="2"
    ry="2"
    width="18"
    x="3"
    y="3"
  />
  <line
    x1="3"
    x2="21"
    y1="9"
    y2="9"
  />
  <line
    x1="3"
    x2="21"
    y1="15"
    y2="15"
  />
  <line
    x1="9"
    x2="9"
    y1="3"
    y2="21"
  />
  <line
    x1="15"
    x2="15"
    y1="3"
    y2="21"
  />
</svg>
`;
